runtime/callback.c

```diff
@@ -386,32 +386,34 @@ static unsigned int hash_value_name(char const *name)
 
 CAMLprim value caml_register_named_value(value vname, value val)
 {
+  CAMLparam2(vname, val);
   const char * name = String_val(vname);
+  size_t namelen = strlen(name);
   unsigned int h = hash_value_name(name);
   int found = 0;
 
   caml_plat_lock_non_blocking(&named_value_lock);
+  name = NULL; /* block may have moved while we waited for the lock. */
   for (struct named_value *nv = named_value_table[h];
        nv != NULL;
        nv = nv->next) {
-    if (strcmp(name, nv->name) == 0) {
+    if (strcmp(String_val(vname), nv->name) == 0) {
       caml_modify_generational_global_root(&nv->val, val);
       found = 1;
       break;
     }
   }
   if (!found) {
-    size_t namelen = strlen(name) + 1;
     struct named_value * nv =
       caml_stat_alloc(sizeof(struct named_value) + namelen);
```

Comment:
On the line here, I should it should be `namelen + 1`.

Could we maybe move the `size_t namelen` declaration back here, to make the patch less invasive? Also there is no need to compute the length in the `found` case.

Reply:
Looking at this a bit more, I am not sure that this is the issue causing the CI failure. The patch here does change the semantics, but a previous version of the runtime did use `sizeof(struct named_value) + strlen(name)` and it worked fine. (The janestreet fork still has that version; trunk changed this in https://github.com/ocaml/ocaml/pull/13354/files#diff-d0b865965fecc49950af30145c97a8a4d79decea131868bdee82c228b32c2404R407 ).

Reply:
Aha! At ocaml-flambda/flambda-backend, we still have [size-1 arrays](https://github.com/ocaml-flambda/flambda-backend/blob/0c7e618ede64a624c550f8ad3f74d9c7ac5655b1/runtime/callback.c#L427) for stretchy trailing array elements in structures like `struct named_value`. So yes, this needs a +1 for the `caml_stat_alloc`.